src/mpi/coll/allreduce/allreduce_intra_recursive_multiplying.c

```diff
@@ -0,0 +1,202 @@
+                    mpi_errno = MPIC_Isend(recvbuf, count, datatype, dst, MPIR_ALLREDUCE_TAG, comm_ptr,
+                                        &reqs[num_reqs++], errflag);
+                    MPIR_ERR_CHECK(mpi_errno);
+                    mpi_errno = MPIC_Irecv(((char *)tmp_buf) + exchanges * count * extent, count, datatype, dst, MPIR_ALLREDUCE_TAG, comm_ptr,
+                                        &reqs[num_reqs++]);
+                    MPIR_ERR_CHECK(mpi_errno);
+                    exchanges++;
+                }
+            }
+
+            /* Wait for asynchronous operations to complete */ 
+            MPIC_Waitall(num_reqs, reqs, MPI_STATUSES_IGNORE);
+            num_reqs = 0;
+            exchanges = 0;
+
+            /* Perform reduction on the received values */
+            int recvbuf_last = 0;
+            for(int dst = rank_offset; dst < starting_rank + next_distance - distance; dst += distance) {
+                void *dst_buf = ((char *)tmp_buf) + exchanges * count * extent;
+                if(dst == rank - distance) {
+                    mpi_errno = MPIR_Reduce_local(dst_buf, recvbuf, count, datatype, op);
+                    MPIR_ERR_CHECK(mpi_errno);
+                    recvbuf_last = 1;
+                    exchanges++;
+                }
+                else if(dst == rank){
+                    mpi_errno = MPIR_Reduce_local(recvbuf, dst_buf, count, datatype, op);
+                    MPIR_ERR_CHECK(mpi_errno);
+                    recvbuf_last = 0;
+                }
+                else {
+                    mpi_errno = MPIR_Reduce_local(dst_buf, (char *)dst_buf + count * extent, count, datatype, op);
+                    MPIR_ERR_CHECK(mpi_errno);
+                    recvbuf_last = 0;
+                    exchanges++;
+                }
+            }
+            
+            if(!recvbuf_last) {
+                mpi_errno = MPIR_Localcopy((char *)tmp_buf + exchanges * count * extent, count, datatype, recvbuf, count, datatype);
+                MPIR_ERR_CHECK(mpi_errno);
+            }
+
+            distance = next_distance;
+            next_distance *= k;
+            exchanges = 0;
+        }
+    }
+
+    /* POST-STAGE: Send result to ranks outside main algorithm */
+    if(pofk < comm_size) {
+        if(rank >= pofk) {
+            int post_src = pofk - (rank % pofk) - 1;
+            /* This process is outside the core algorithm, so receive data */
+            mpi_errno = MPIC_Recv(recvbuf, count,
+                                  datatype, post_src,
+                                  MPIR_ALLREDUCE_TAG, comm_ptr, MPI_STATUS_IGNORE);
+            MPIR_ERR_CHECK(mpi_errno);
+        } else {
+            /* This is process is in the algorithm, so send data */
+            for (int post_dst = 2 * pofk - (rank % pofk) - 1; post_dst < comm_size; post_dst += pofk) {
+                mpi_errno = MPIC_Send(recvbuf, count, datatype, post_dst, MPIR_ALLREDUCE_TAG, comm_ptr,
```

Comment:
Use Isend + Waitall for better bandwidth.

Reply:
Fixed!